File: Data Structures & Algorithms/interleaving-string/submission-7.cpp
class Solution {
public:
    vector<vector<vector<int>>> memo;
    bool dfs(int i,int j , int k,string s1, string s2, string s3){
        if(k==s3.size())
           return i==s1.size() && j==s2.size();
        if(memo[i][j][k]!=-1)
          return memo[i][j][k];
        if(i<s1.size() && s1[i]==s3[k])
           if(dfs(i+1,j,k+1,s1,s2,s3)){
               memo[i][j][k]=1;
               return true;
           }
        if(j<s2.size() && s2[j]==s3[k])
           if(dfs(i,j+1,k+1,s1,s2,s3)){
               memo[i][j][k]=1;
               return true;
           }
        memo[i][j][k]=0;
        return false;
        
    }
    bool isInterleave(string s1, string s2, string s3) {
        memo.assign(  s1.size() + 1,vector<vector<int>>(s2.size() + 1,vector<int>(s3.size() + 1, -1)));
        return dfs(0,0,0,s1,s2,s3);
    }
};

File: Data Structures & Algorithms/target-sum/submission-7.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number}
     */
    findTargetSumWays(nums, target) {
        const n = nums.length;
        const memo = new Map();
        const dfs = (i,cur)=>{
            const key = i + "," + cur;
            if(memo.has(key)) 
                return memo.get(key);
            if(i==n-1)
               return cur===target ? 1 : 0;
            memo.set(key,dfs(i+1,cur+nums[i+1])+dfs(i+1,cur-nums[i+1]));
            return memo.get(key);
        }
        return dfs(0,nums[0])+dfs(0,-nums[0]);
    }
}
